jmh-benchmarks/src/main/java/org/apache/kafka/jmh/assignor/ServerSideAssignorBenchmark.java

```diff
@@ -0,0 +1,252 @@
+
+        for (int i = 0; i < topicCount; i++) {
+            Uuid topicUuid = Uuid.randomUuid();
+            String topicName = "topic" + i;
+            allTopicIds.add(topicUuid);
+            topicMetadata.put(topicUuid, new TopicMetadata(
+                topicUuid,
+                topicName,
+                partitionsPerTopicCount,
+                partitionRacks
+            ));
+        }
+
+        return topicMetadata;
+    }
+
+    private void createAssignmentSpec() {
+        Map<String, AssignmentMemberSpec> members = new HashMap<>();
+
+        int topicCounter = 0;
+        Map<Integer, Set<Uuid>> tempMemberIndexToSubscriptions = new HashMap<>(memberCount);
+
+        // In the rebalance case, we will add the last member as a trigger.
+        // This is done to keep the total members count consistent with the input.
+        int numberOfMembers = simulateRebalanceTrigger ? memberCount -1 : memberCount;
+
+        for (int i = 0; i < numberOfMembers; i++) {
+            if (subscriptionModel == SubscriptionModel.HOMOGENEOUS) {
+                addMemberSpec(members, i, new HashSet<>(allTopicIds));
+            } else {
+                Set<Uuid> subscribedTopics = new HashSet<>(Arrays.asList(
+                    allTopicIds.get(i % topicCount),
+                    allTopicIds.get((i+1) % topicCount)
+                ));
+                topicCounter = max (topicCounter, ((i+1) % topicCount));
+                tempMemberIndexToSubscriptions.put(i, subscribedTopics);
+            }
+        }
+
+        int lastAssignedTopicIndex = topicCounter;
+        tempMemberIndexToSubscriptions.forEach((memberIndex, subscriptions) -> {
+            if (lastAssignedTopicIndex < topicCount - 1) {
+                subscriptions.addAll(allTopicIds.subList(lastAssignedTopicIndex + 1, topicCount - 1));
+            }
+            addMemberSpec(members, memberIndex, subscriptions);
+        });
+
+        this.assignmentSpec = new AssignmentSpec(members);
+    }
+
+    private Optional<String> rackId(int index) {
+        return isRackAware ? Optional.of("rack" + index % NUMBER_OF_RACKS) : Optional.empty();
+    }
+
+    private void addMemberSpec(
+        Map<String, AssignmentMemberSpec> members,
+        int memberIndex,
+        Set<Uuid> subscribedTopicIds
+    ) {
+        String memberName = "member" + memberIndex;
+        Optional<String> rackId = rackId(memberIndex);
+
+        members.put(memberName, new AssignmentMemberSpec(
+            Optional.empty(),
+            rackId,
+            subscribedTopicIds,
+            Collections.emptyMap()
+        ));
+    }
+
+    private static Map<Integer, Set<String>> mkMapOfPartitionRacks(int numPartitions) {
+        Map<Integer, Set<String>> partitionRacks = new HashMap<>(numPartitions);
+        for (int i = 0; i < numPartitions; i++) {
+            partitionRacks.put(i, new HashSet<>(Arrays.asList("rack" + i % NUMBER_OF_RACKS, "rack" + (i + 1) % NUMBER_OF_RACKS)));
```

Comment:
In practice, partitions will have three racks. Should we have three here?

Reply:
I wanted to test with less options for each rack match, otherwise most of the partitions can be assigned to a lot of member. It is also easier to manually check if the algo is correct, I can change it if it's necessary though

Reply:
I see your point. This is totally valid in tests. However, I wonder if it is representative in benchmarks.

Reply:
okay I will just change it

Reply:
In practice, are brokers of a cluster spread across 3 racks as well or can there be more?

Reply:
Otherwise every partition is basically going to be available in every rack barring any bugs in the replica distribution or broker distribution